### ISSUE TITLE
Take advantage of separate compilation

### ISSUE THREAD

**Issue by eschaton** [simh/simh#697]
Building all of SIMH takes longer than it should because each simulator build recompiles every `.c` file, even the ones that don't change between simulators.

The SIMH makefile should be adjusted to build each `.c` file into a `.o` file and to build the simulators by linking the `.o` files. Doing so can significantly speed up the build process, and also make it easier to hack on SIMH.

**Comment by markpizz** [comment]
That's a nice complaint.  What you're suggesting doesn't speed up the build process at all when starting from scratch (the most common case).  Meanwhile, on my 18 month old desktop, the largest simulator takes 30 seconds to compile.  Yes it could be less, but that's fine with me.

If you can come up with an alternative approach that's somewhat easy to adopt that sanely allows building of all of the simulators (which each have to be compiled with DIFFERENT -D options, even on the shared modules) we'll be glad to take a closer look.

**Comment by eschaton** [comment]
If the “shared” modules have to be compiled with *different* `-D` options then they aren’t really shared, are they? Instead of parameterizing this code at compile time it should probably be parameterized at runtime.

**Comment by markpizz** [comment]
Yes, they are NOT shared in the shared library sense, nor did I suggest they were.  Many source modules happen to be used in all (or most) simulators.  Meanwhile, the average simh user is absolutely only interested in one or at most a couple of simulators so any specific benefits of sharing aren't a great concern.

It's really easy to be critical without actually contributing anything but criticism.  We'll welcome useful contributions, but please stop just throwing stones.

**Comment by eschaton** [comment]
@markpizz Filing bug reports and enhancement requests is not “throwing stones,” and neither is pointing out things that can be improved. In my career, I’ve seen that reacting to feedback in this way just serves to dissuade others from providing feedback, rather than raise the level of contribution from those in the community. 

Furthermore, for many people, this kind of basic feedback may be the most they can reasonably contribute because of complications like employment agreement IP assignments. For example, if I made the changes I suggest myself (which I am quite capable of, thank you), I would not be able to share them with you without engaging a bureaucracy, and probably not then either.

**Comment by sethm** [comment]
I would like to add my vote for this feature request as well, though I realize it is not a high priority.

During development of the 3B2 simulator, I wrote a very small, non-portable Makefile for my own use which did compilation of each unit independently before linking them. On my primary Linux development machine, the default makefile takes 22.9 seconds to compile and link the 3B2 simulator. In contrast, the custom one-off makefile takes 5.2 seconds by using all available cores.

I fully acknowledge this is not an issue anybody but a simulator developer will ever care about, but when doing rapid code, compile, test cycles, it made a big difference for me personally. I also fully acknowledge that it's not a trivial feature request and would require significant effort.

**Comment by rcornwell** [comment]
I will also add my vote. However I have looked at doing this several times and it is not as easy as it sounds. At a minimum there would need to be at least two libraries, one for 32 bit value_t and one for 64 bit value_t. I am not sure how much else would be effected by conditional compilation in the various modules. For me a complete rebuild of my simulators takes a minute or three so it is not that bad. Fully testing my simulators now takes about 1 day.

It is also unclear how this would work on Windows or VMS (if it is still a supported platform).

Unlike other platforms out there for writing simulators, SimH is more a collection of related simulators sharing a bunch of common functions, rather then a unified group of simulators. Each of them uses the features that they need from simH and leaves those they don't untouched.

If you believe you can implement this, then I think we would all like to see how it is done. However since most people only build one or two simulators that they are interested in, I am not sure how much use this would be.

**Comment by Rhialto** [comment]
One advantage of the current all-at-once compilation is perhaps that the
compiler can more easily do global optimizations, that are not available
when simh is compiled piecemeal. (Although some compilers support
storing sufficient information in the object files to do link-time
optimization anyway)

**Comment by 9track** [comment]
If I need to build more than one simulator at once then I specify the jobs flag to make to speed things up.
```
make -j14 vax730 vax750 vax780 vax8200 ...
```
This will build the simulators in parallel taking advantage of multiple CPU cores.

**Comment by AK6DN** [comment]
A question on 9track's use of the -j flag to run multiple compile targets in parallel ... can this not result in collisions where the same source module needs to be compiled with two different -D flags? So could a race condition result where a simulator gets built with the wrong -D option?  Just asking.

**Comment by Rhialto** [comment]
The `-j14` flag makes make run up to 14 commands in parallel, in our
case all compilations I suppose. They are all independent from each
other, just like compilations you would start from different shells.
They each use their own working files. It would be a mess otherwise :-)

**Comment by AK6DN** [comment]
Well I ask because the same source .c file is used, I assume compilation with different -D options would still place the compiled .o files in the same directyory, which would result in a conflict. If I launched several makes from different login windows in the same directory that could result in a mess ... same source file wanting to be compiled with different -D options into .o file(s) in the same directory. How is -jNN different? Unless it tags the generated .o files with a job id I don't see how it does NOT conflict?

**Comment by AK6DN** [comment]
All this being said, I still don't see the justification for this added complexity. The makefiles already are horrendously complex supporting a wide range of simulators, options, and environments (eg, I use cygwin). I just see the ROI.

**Comment by 9track** [comment]
I suppose it might depend on your compiler but when I build with GCC the object files actually go to a temporary directory using a randomly generated filename (because the compilation and linking is done in a single step with the current makefile) so there will be no conflict.

**Comment by AK6DN** [comment]
If that is the case (TBH I was working off old memories where .c => .o => .exe) so if that is the case now my knowledge is obsolete and I respectfully withdraw the comment. Thanks!

**Comment by eschaton** [comment]
Just for some further background, I've been putting together an Xcode workspace for SIMH and separate compilation works beautifully, it builds almost instantly on even underpowered hardware like a MacBook. I have one "libsimh" target that produces libsimh_a32i32.a and libsimh_a64i64.a, and then I have projects that build a couple of the simulators, and a couple of xcconfig files for common settings. They work together nicely and make it *very* fast to build and debug simulators.

**Comment by markpizz** [comment]
The simh makefile build mechanisms in the current https://github.com/simh/simh repo can now build simulators while compiling source modules one at a time.  The default is unchanged, that is each simulator is build with a single compile activation.

Invoking make with BUILD_SEPARATE=1 on the command line (or if BUILD_SEPARATE=1 is an exported environment variable), will cause separate source modules to be built as needed.
